Deploy and sign off (#96)
Clinical-Genomics/development: docs/publish/sign-off.md

# When do you need a sign-off and who can sign off on tests, code reviews and deploys?

> All tools developed at clinical genomics needs to be tested, reviewed and deployed according to a specific routine which is generally outlined under ["How to publish your tool"](prod.md). These three things also needs a sign-off by an authorized person. For tools used in production this is especially important. There are two development models used at Clinical Genomics and depending on which you use the sign-offs happen at slightly different steps in the update process.

Reviewer: ```suggestion
All tools developed at clinical genomics need to be tested, reviewed and deployed according to a specific routine which is generally outlined under ["How to publish your tool"](prod.md). These three things also need a sign-off by an authorized person. For tools used in production this is especially important. There are two development models used at Clinical Genomics and depending on which you use the sign-offs happen at slightly different steps in the update process.
```

Author: I've changed that

## Github flow

> In this model the sign-offs all happen in the pull request (PR). The developer will as part of the initial PR add two check boxes to the description; one for code review and one for functional testing. The authorized person(s) doing the test and review then ticks the boxes as the code has been checked and tested on stage according to the test case described in the PR. Only when the boxes are ticked, the authorized person(s) have added their thumbs up and the PR has been approved the decision to merge and deploy is made. The person(s) authorized to sign-off on this then adds the go-ahead as a comment in the PR and merges the branch to master (or other source). This process is better described in ["How to publish your tool"](prod.md).

Reviewer 2: You could use the following style to point to the link you want to include: `["How to publish your yool"][prod]` and then at the end of the document add `[prod]: prod.md`. This way, everytime you mention `[prod]`, the reference can be changed in only one place.

Author: Aha! Smart!

# When do you need a sign-off and who can sign off on tests, code reviews and deploys?

> All tools developed at clinical genomics needs to be tested, reviewed and deployed according to a specific routine which is generally outlined under ["How to publish your tool"][prod]. These three things also needs a sign-off by an authorized person. For tools used in production this is especially important. There are two development models used at Clinical Genomics and depending on which you use the sign-offs happen at slightly different steps in the update process.

Author: ```suggestion
All tools developed at clinical genomics need to be tested, reviewed and deployed according to a specific routine which is generally outlined under ["How to publish your tool"][prod]. These three things also need a sign-off by an authorized person. For tools used in production this is especially important. There are two development models used at Clinical Genomics and depending on which you use the sign-offs happen at slightly different steps in the update process.
```

## Github flow

> In this model the sign-offs all happen in the pull request. The developer will as part of the initial pull request add two check boxes to the description; one for code review and one for functional testing. The authorized person(s) doing the test and review then ticks the boxes as the code has been checked and tested on stage according to the test case described in the pull request. Only when the boxes are ticked, the authorized person(s) have added their thumbs up and the pull request has been approved the decision to merge and deploy is made. The person(s) authorized to sign-off on this then adds the go-ahead as a comment in the pull request and merges the branch to master (or other source). This process is better described in ["How to publish your tool"][prod].

Reviewer 2: Just thought of something: most repos are configured so that you cannot merge unless you get an approving review. Last one who ticks, should also approve.

Reviewer 2: We.might need a 3rd party app that can give us two types of pull request approval: pullapprove.com

Author: Pullapprove looks quite useful. But let's look into that at a later stage.

I'll add the approval on last tick.

### People authorized to sign-off on code reviews
* Code owner(s) of the specific repository

### People authorized to sign-off on functional tests
* Clinical Genomics employee

### People authorized to sign-off on merge and deploy
* Tools in production - Team leader IT (see AM document `1047 Functions and signature list`)
* Tools in development - Code owner(s) of the specific repository

## Git flow
In this model several pull requests are bunched in the same release and it is therefore necessary to do the sign-offs in three steps:

1. First pull request to add changes to the feature branch - sign-offs needed are on code review, functional testing and merge as described under Github flow.

> 2. AM Test specification and implementation plan - sign-offs needed on system testing (testing of the whole feature branch to be deployed) and decision to merge to master and deploy the code. Test specifications and implementation plan templates are available in AM (`1249 Test Specification - Template` and `1187 Implementation plan - template`)

Reviewer 2: ```suggestion
2. AM Test specification and implementation plan - sign-offs needed on system testing (testing of the whole development or release branch to be deployed) and decision to merge to master and deploy the code. Test specifications and implementation plan templates are available in AM (`1249 Test Specification - Template` and `1187 Implementation plan - template`)
```
